### PR TITLE
AP-78 Service for letter writing

### PR DESCRIPTION
**Pull request by vwvw**
Test failing as in devel

### CODE REVIEW

File: mobile_app_connector/models/compassion_correspondence.py [hunk 1]
```diff
@@ -114,3 +115,70 @@ def _get_required_param(self, key, params):
         if key not in params:
             raise ValueError('Required parameter {}'.format(key))
         return params[key]
+
+    def mobile_get_preview(self, *_, **other_params):
+        body = self._get_required_param('letter-copy', other_params)
```
Comment (ecino) [inline mobile_app_connector/models/compassion_correspondence.py:138]: A method comment explaining the function would be welcome. Please rename as well `_` parameter which interfers with the translation function of Odoo. `args` is more pythonic.

File: mobile_app_connector/models/compassion_correspondence.py [hunk 2]
```diff
@@ -114,3 +115,70 @@ def _get_required_param(self, key, params):
...
+    @api.multi
+    def mobile_send_letter(self, *params, **other_params):
+        params = params[0]
+        template_id = self._get_required_param('TemplateID', params)
+        if 'DbId' in params:
+            # The letter was submitted on the last api call
+            if template_id == "0":
+                return "Letter Submitted "
+            return "Card Submitted "
+        # body = self._get_required_param('Message', params)
```
Comment (ecino) [inline mobile_app_connector/models/compassion_correspondence.py:196]: Cleanup your code

File: mobile_app_connector/models/compassion_correspondence.py [hunk 3]
```diff
@@ -114,3 +115,70 @@ def _get_required_param(self, key, params):
...
+        gen.preview()
+        web_base_url = \
+            self.env['ir.config_parameter'].get_param('web.external.url')
+        url = web_base_url + "/web/image/" + gen._name + "/" + \
+            str(gen.id) + "/preview_pdf"
+        return url
+
+    @api.multi
+    def mobile_send_letter(self, *params, **other_params):
+        params = params[0]
```
Comment (ecino) [inline mobile_app_connector/models/compassion_correspondence.py:189]: A method comment would be welcome.

File: mobile_app_connector/models/compassion_correspondence.py [hunk 4]
```diff
@@ -114,3 +115,70 @@ def _get_required_param(self, key, params):
...
+        # partner_id = self._get_required_param('SupporterId', params)
+        # iOS and Android do not return the same format
+        if template_id == '0' or template_id == 0:
+            # write a card -> default template
+            template_id = '2'
+        child_id = self._get_required_param('Need', params)
+        if isinstance(child_id, list):
+            child_id = child_id[0]
+        child = \
+            self.env['compassion.child'].search([('id', '=', str(child_id))])
```
Comment (ecino) [inline mobile_app_connector/models/compassion_correspondence.py]: Is id an integer? why converting it to string? I think this is more appropriate :
```suggestion
            self.env['compassion.child'].browse(child_id)
```

File: mobile_app_connector/models/mobile_writing.py [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+    external_url = fields.Char(translate=True)
+    date_start = fields.Date(
+        readonly=True,
+        states={'new': [('readonly', False)]}
+    )
+    date_stop = fields.Date(
+        readonly=True,
+        states={'new': [('readonly', False)]}
+    )
+    is_active = fields.Boolean()
```
Comment (ecino) [inline mobile_app_connector/models/mobile_writing.py]:
```suggestion
    active = fields.Boolean()
```
We did it badly in other models, but `active` is a standard Odoo field which is there for that.

File: mobile_app_connector/models/mobile_writing.py [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+        return [
+            x.get_json() for x in actives
+        ]
+
+    @api.multi
+    def get_json(self):
+        self.ensure_one()
+        web_base_url = \
+            self.env['ir.config_parameter'].get_param('web.external.url')
+        base_url = web_base_url + "/web/image/" + self.template_ids[0]._name \
```
Comment (ecino) [inline mobile_app_connector/models/mobile_writing.py:109]: Why having a many2many relation if we use only the first relation? Wouldn't be better to change it to a many2one field?

**Review by ecino — CHANGES_REQUESTED** [review]

**Comment by vwvw** [comment]
One minor fix coming

**Review by ecino — APPROVED** [review]

**Comment by ecino** [comment]
Ok, waiting for it.

**Pull request merged**